dvc/data/stage.py

```diff
@@ -247,27 +250,117 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
-def _load_from_state(odb, staging, fs_path, fs, name):
+def _load_raw_dir_hash(
+    odb: "ObjectDB",
+    staging: "ReferenceObjectDB",
+    fs_path: "AnyFSPath",
+    fs: "FileSystem",
+    hash_info: "HashInfo",
+    meta: "Meta",
+    name: "str",
+    dry_run: bool,
+) -> "Tree":
+    from dvc.fs._callback import NoOpCallback
+    from dvc.ui import ui
+
+    from .tree import Tree
+
+    tree = _try_load(odb, hash_info.as_raw(), raw_dir_hash=True)
+    if tree is None:
+        raise FileNotFoundError
+
+    assert isinstance(tree, Tree)
+    assert tree.fs
+    meta.nfiles = len(tree)
+    assert tree.hash_info.name == name
+
+    if not dry_run:
+        with ui.status(f"Loading {fs_path}"):
+            for key, _, oid in tree:
+                staging.add(
```

Comment:
Just a quick note about this: since we [recently](https://github.com/iterative/dvc/commit/5067e1f5bc3335431c50b670282b1c85f11f3d3e) modified `ObjectDB.add` with callbacks
https://github.com/iterative/dvc/blob/b14a972cd057a5fcf6dfbee5da475e30c1f36bec/dvc/objects/db.py#L133-L146
calling `add()` with no callback arguments results in a `Tqdm` instance being created for each call, which results in a massive slow down when adding a large amount of files (at least 4x in my limited testing).

A quick solution is to modify this call with:
```python
with NoOpCallback() as cb:
    for key, _, oid in tree:
        staging.add(
            ...,
            callback=cb
        )
```
Note that this won't affect current behaviour, because the callback is never called for `ReferenceObjectDB._add_file` (it is only called for a directory, since that calls `ObjectDB._add_file`). 
Instantiating `Tqdm` prints a progressbar for the given file but this is never updated (and is too quick to be of any use anyway).

If we wish to show a progressbar for the state of staging, we'd also have to add something like:
```python
with NoOpCallback() as cb:
    with Tqdm(
        unit="files",
        desc=f"Loading {odb.fs.path.name(fs_path)}",
        total=meta.nfiles + 1,
    ) as pbar:
        for key, _, oid in tree:
            staging.add(
                ...,
                callback=cb
            )
             pbar.update()
```

Reply:
Note that this also makes `transfer` very slow, since it calls `dest.add` with no callbacks.

Reply:
Just for context, for a directory with ~10k files:

- no `callback` kwarg to `staging.add`: `dvc status  6.52s user 0.41s system 98% cpu 7.017 total`
- with `callback=noopcallback`: `dvc status  2.26s user 0.14s system 99% cpu 2.410 total`

cc @skshetry

Reply:
@dtrifiro, I think it makes sense to use `fs._callback.DEFAULT_CALLBACK` in the case of individual objects, and use Tqdm one for the tree. (My understanding is that they are just saving to configs to `refodb`/`memfs` right?)